1027 historic charges

entantoencuanto opened the pull request:
Closes PopulateTools/issues#1027
Closes PopulateTools/issues#1029
Closes https://github.com/PopulateTools/issues/issues/1110
Related to PopulateTools/issues#1031

## :v: What does this PR do?

* Adds a new model `GobiertoPeople::Charge` as described in the issue
* The charge has a `start_date` and an `end_date`. nil values are accepted: A nil value means that the interval has no limit for the corresponding boundary. For example, a charge with blank `start_date` and `end_date` implies that is in effect for all the time
* Adds new associations of gifts, invitations, trips and events with departments by using a concern. This allows us for gifts, invitations, trips or events to get the charge and department of the person in the date or start date of the resource
* Adds associations to departments with these resources to get the items of a department (all of them belongs to a person and this person has a charge in the department in a period in which the date or start date is included)
* Changes the `charge` method of people to accept a date and obtain the charge among the charges of the person compatible with the date. If no date is provided, the charge is obtained from the current date or if there is no charge at current date the last charge. If there are no charges created for the person the `charge` attribute of the model is taken (in a future this field probably will be removed)
* For pages which allows the selection of an interval of dates, the charge used currently for the people is the charge at the end of the interval, but it has been added to the person serializer new attributes with the charges of the person in the interval ordered by date. The charges for each user can be passed to the serializer from the controller to save requests to the database
* For events, the charge associated with it and displayed is the charge at the beginning of the event
* Adds some fixtures for charges and changes some tests with the new fixtures scenarios.
* Fixes some queries used by the custom engine to take into account events related with people belonging to a department via charge.
* I've tried to put the charge associated to each item in the lists and person page but since the names are sometimes very long it looks weird.
* Generates autolinks with the URLs present in the trips descriptions

## :mag: How should this be manually tested?

For the moment it can't be tested in staging, so it should be done locally


## :shipit: Does this PR changes any configuration file?

No

(Changes in these files might need to update the role in Ansible)

## :book: Does this PR require updating the documentation?

No

entantoencuanto committed:
Add migration to create gp_charges table

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
Add GobiertoPeople::Charge model

entantoencuanto committed:
Define concern to include associations with people via charge to models

entantoencuanto committed:
Add BelongsToPersonWithCharge to GobiertoPeople::Gift model

entantoencuanto committed:
Add charges tests fixtures

entantoencuanto committed:
WIP for events, invitations and trips

entantoencuanto committed:
Unify names for date intervals boundaries

entantoencuanto committed:
Use associations with charges in department decorator

entantoencuanto committed:
Move date_range_sql to Charge model

entantoencuanto committed:
Add scopes to charge

entantoencuanto committed:
Include positions list in person serializer

entantoencuanto committed:
Interpolate table_name in sql literals

entantoencuanto committed:
Update tests

entantoencuanto committed:
Replace charge Person method accepting a date and use date when possible

entantoencuanto committed:
Fix test

entantoencuanto committed:
Fix test

entantoencuanto committed:
Add migration to include external_id in charges

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
Fix method and add external_id concern to charges model

entantoencuanto committed:
Change people serializer to include charges and exclude not required associations

entantoencuanto committed:
Avoid not_null condition on older events association with departments

entantoencuanto committed:
Allow rowchart_items_query to return items with 0 counters

entantoencuanto committed:
Add query for people with activities

entantoencuanto committed:
Fix date_range_sql to work with string keys

entantoencuanto committed:
Revert join

entantoencuanto committed:
Change person serializer to provide position data with filtered charges

entantoencuanto committed:
Allow requests to people API not filtering people positions between dates

entantoencuanto committed:
Add tipsy in people view

entantoencuanto committed:
Fix tests

entantoencuanto committed:
Add scope to take into account event_attendees with department associated in custom engine data

entantoencuanto committed:
Fix tests

ferblape commented on app/models/gobierto_people/person.rb:
```diff
@@ -64,6 +65,24 @@ def owned_attending_events
+      else
+        return if charge_translations.blank?
+
+        charge_translations.fetch(I18n.locale, charge_translations&.values&.first)
```
If `charge_translations` is a regular translate attribute why not replace the whole `else` block by a plain `super`? Would it work?

entantoencuanto replied:
Actually `charge` is overrided but it's not given by a superclass, is introduced with `translates` call in the model, calling super in this context generates an error saying that the method is not defined in any superclass. But if we create a superclass with the `translate  :charge` statement and the method is inherited from it, the `super` call raises an exception because the charges superclass doesn't accept arguments and this accepts an optional argument.

ferblape commented on app/queries/gobierto_people/people_with_activities_query.rb:
```diff
@@ -0,0 +1,106 @@
+    def initialize(params = {})
+      @site = params[:site]
+      @conditions = params[:conditions].symbolize_keys
+      @relation = (params[:relation] || model.all).where(id: people_with_activities.map(&:id)).left_outer_joins(events_association)
```
Do we need the `all`?

entantoencuanto replied:
No, hehe.

ferblape commented on app/queries/gobierto_people/people_with_activities_query.rb:
```diff
@@ -0,0 +1,106 @@
+        "#{table_name}.id = #{association_model}_relation.id"
+      end.join(" OR ")
+
+      "WITH #{cte} SELECT DISTINCT(#{table_name}.*) FROM #{table_name} #{joins} WHERE #{where}"
```
Could we replace the `.* ` by `.id`? This query is only used here and you are only using the ids of the people

entantoencuanto replied:
Changed

ferblape commented on app/queries/gobierto_people/people_with_activities_query.rb:
```diff
@@ -0,0 +1,106 @@
+    end
+
+    def activities_sql
+      cte = ASSOCIATIONS.keys.map do |association_model|
```
What does cte stands for? The variable name could be improved a bit 😝

entantoencuanto replied:
It stands for `common_table_expression`. I've changed `cte` with it, it's much more expressive ;)

ferblape commented on app/queries/gobierto_people/people_with_activities_query.rb:
```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
```
I've been reviewing this class during at least 10 minutes and I don't fully follow the queries you are generating. I think you could include some comments with the queries the method `activities_sql` generates

entantoencuanto replied:
I've added some comments

ferblape commented on app/serializers/gobierto_people/person_serializer.rb:
```diff
@@ -28,11 +34,43 @@ def political_group
+      object.historical_charges.reverse_sorted.join("<br>")
+    end
+
+    def filtered_positions
```
memoize please, it's used in many methods

entantoencuanto replied:
Totally agree.

ferblape commented on app/views/gobierto_people/people/_person.html.erb:
```diff
@@ -16,7 +16,9 @@
       <h2><%= person.name %></h2>
     <% end %>
-    <div class="soft"><%= person.charge %></div>
+    <div class="soft tipsit" title="<%= person.historical_charges.reverse_sorted.join("<br>") %>">
```
- This is a N + 1, are you including the historical_charges relation when loading the person?
- could you do a helper or move it to a presenter?

entantoencuanto replied:
Done

ferblape reviewed (changes requested):
Looks great, but it could be improved a bit ;)

entantoencuanto committed:
Add method to concern to return name and historical charge

entantoencuanto committed:
Remove useless .all

entantoencuanto committed:
Improve variable name and query

entantoencuanto committed:
Add comments

entantoencuanto committed:
Remove unused serialization attribute

entantoencuanto committed:
Memoize method

entantoencuanto committed:
Add historical_charges_names to PersonDecorator

entantoencuanto committed:
Use decorated collections and eager load of historical_charges for people in controllers and views

entantoencuanto committed:
Use local_assign if present in people partial

ferblape commented:
@entantoencuanto can you check the conflicts?

entantoencuanto committed:
tMerge branch 'master' into 1027-historic_charges

* master: (96 commits)
  Changing isNan to Number.isNan
  Fixing the same potential issue with NaN amounts in contracts controller
  Amend character escaped char
  Update aws-sdk-s3 to version 1.73.0
  Check value of amount is NaN
  Fix form for text custom fields with translations
  resolved changes @crashillo
  remove template
  Rescue flaky test exception
  Update i18n-js to version 3.7.1
  Use cookie locale only if available on current site
  Add integration test
  Update cloudinary to version 1.16.0
  Update google-api-client to version 0.41.1
  Remove fingerprint from assets paths in processes fixtures
  Update aws-sdk-s3 to version 1.72.0
  Allow sending arbitrary json for visualizations spec attribute
  Update visualization test with a spec including nil elements in arrays
  restore row_pivots, column_pivots when load a visualization | refactor toggleConfig
  Update simple_calendar to version 2.4.1
  ...

entantoencuanto committed:
Change oputput of charges date_range

entantoencuanto committed:
Merge branch 'master' into 1027-historic_charges

* master: (542 commits)
  Update all of rails to version 6.0.3.3
  Remove watched_paths from engines
  Update paper_trail to version 11.0.0
  Capture exception on meta serializer if dataset table can't be accessed
  Update redis to version 4.2.2
  Update codecov to version 0.2.11
  Update google-api-client to version 0.44.1
  Update puma to version 4.3.6
  Update codecov to version 0.2.10
  Bump http-proxy from 1.18.0 to 1.18.1
  Update bcrypt to version 3.1.16
  Update aws-ses to version 0.7.0
  Bump node-sass from 4.12.0 to 4.14.1
  Fix missing deprecated access to config_for hash using string keys
  Update invisible_captcha to version 1.1.0
  Add missing association in BudgetLineFeedback validation test
  Fix associations in Verification::IdNumber model test
  Add association to event_attendees to ensure validity on event form
  Relax GobiertoCalendars::EventLocation association with events
  Correct fixture
  ...

Crashillo committed:
Update person_serializer.rb

entantoencuanto committed:
Allow API people index to return people with any activity between dates

entantoencuanto committed:
Change counter in welcome home statistics

entantoencuanto committed:
Add some tests

ferblape committed:
Add missing import of tipsy

ferblape commented:
tipsy wasn't imported in people module and it was causing a JS error in staging, it should be fixed now

entantoencuanto committed:
Merge branch 'master' into 1027-historic_charges

* master: (460 commits)
  Update google-api-client to version 0.51.0
  Update active_model_serializers to version 0.10.11
  Update rubocop to version 1.5.2
  line-height unitless
  Don't change order of people in people index
  Remove limit of people listed in welcome controller
  renaming
  Update module-TEMP-maker.scss
  replace properly variables
  remove CSS dupes
  Update elasticsearch-extensions to version 0.0.32
  Update aws-sdk-s3 to version 1.86.0
  lil glitches autocomplete
  create autocomplete component
  move text editable to shared components
  widget table bars relative
  Update puma to version 5.1.0
  indicators available
  indicators data from endpoint
  Fix order in routes to avoid considering meta as a dataset slug
  ...

entantoencuanto committed:
Update db/structure.sql

entantoencuanto committed:
Merge branch 'master' into 1027-historic_charges

* master:
  Update rubocop to version 1.6.1
  Fix velocity missing require
  Bump ini from 1.3.5 to 1.3.7
  Only redirect the specific paths, avoiding the redirection with urls like /api/v1/dashboards
  fix locales
  name async
  set translations

entantoencuanto committed:
Merge branch 'master' into 1027-historic_charges

* master: (58 commits)
  Update i18n-tasks to version 0.9.32
  Update webmock to version 3.11.0
  Update elasticsearch-extensions to version 0.0.33
  Update paper_trail to version 11.1.0
  Update paranoia to version 2.4.3
  Update redcarpet to version 3.5.1
  Update appsignal to version 2.11.1
  Fix test to force multipart_form mimetype on request
  rename events by scope
  Don't validate data_file when nil in DatasetForm
  Update ViewerManager.vue
  Update aws-sdk-s3 to version 1.86.2
  integrate dashboard tabs in plans
  Add test to check response when database configuration is missing
  Validate database configuration presence in dataset form
  Use an error message as fallback in write query Connection method
  Update google-api-client to version 0.52.0
  Add some datasets create/update API controller tests
  Validate length of csv_separator in datasets upload
  Capture errors when a string is passed instead of an uploaded file in API create/update dataset
  ...

entantoencuanto committed:
Merge branch 'master' into 1027-historic_charges

* master: (385 commits)
  Update timecop to version 0.9.4
  Update all of rails to version 6.0.3.5
  Remove unused keys
  Fix tests
  dont remove final amount when transform data
  locale ca
  add source to contracts show
  update locales
  Update bootsnap to version 1.7.2
  Update appsignal to version 2.11.6
  Bump marked from 1.2.3 to 2.0.0
  Update timecop to version 0.9.3
  Update Table.vue
  fix routes
  Fix referrer check and refactor
  Update puma to version 5.2.1
  Update codecov to version 0.4.3
  Update bootsnap to version 1.7.1
  sticky column selector
  add domContenLoaded to load responsive menu and relocate legacy modules
  ...
